File: 프로그래머스/2/87946. 피로도/피로도.py
answer = -1
def solution(k, dungeons):
    global visited
    visited = [0] * len(dungeons)
    dfs(k, 0, dungeons, len(dungeons))
    return answer

def dfs(k, cnt, dungeons, N):
    global answer
    if answer < cnt:
        answer = cnt
    for i in range(N):
        if not visited[i] and k >= dungeons[i][0]:
            visited[i] = 1
            dfs(k-dungeons[i][1], cnt+1, dungeons, N)
            visited[i] = 0
            

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;
class Solution {
    int max = 0;

    public int solution(int k, int[][] dungeons) {
        int answer = -1;
        boolean[] visited = new boolean[dungeons.length];
        permutation(k, dungeons, visited, 0);
        return max;
    }
    
    public void permutation(int tired, int[][] dungenons, boolean[] visited, int cnt) { //cnt: 던전 탐험한 수
        
        for (int i = 0; i < dungenons.length; i++) {
            if (tired < dungenons[i][0]) {
                max = Math.max(max, cnt);
            }
            else if (!visited[i]) {
                if (tired >= dungenons[i][0]) {
                    visited[i] = true;
                    permutation(tired-dungenons[i][1], dungenons, visited, cnt + 1);
                    visited[i] = false;
                }   
            }
            
        }
    }
}